brave/src/main/java/brave/propagation/TraceContextOrSamplingFlags.java

```diff
@@ -81,64 +176,130 @@ public TraceContextOrSamplingFlags sampled(@Nullable Boolean sampled) {
+   *   <li>{@link TraceIdContext#sampled()}: {@code null}</li>
+   * </ul>
+   *
+   * @return the trace ID context when {@link #context()} and {@link #samplingFlags()} are not
+   * {@code null}
+   * @see #create(TraceIdContext)
+   * @see #newBuilder(TraceIdContext)
+   * @since 4.9
+   */
   @Nullable public TraceIdContext traceIdContext() {
     return type == 2 ? (TraceIdContext) value : null;
   }
 
+  /**
+   * Returns non-{@code null} when a {@linkplain TraceContext#traceIdString() trace ID} was not
+   * {@link Extractor#extract(Object) extracted} from a request.
+   *
+   * <p>For example, given the header "b3: 1", {@link B3Propagation} extracts {@link #SAMPLED}.
+   *
+   * @return sampling flags when {@link #context()} and {@link #traceIdContext()} are not {@code
+   * null}
+   * @see #create(SamplingFlags)
+   * @see #newBuilder(SamplingFlags)
+   * @since 4.9
+   */
   @Nullable public SamplingFlags samplingFlags() {
     return type == 3 ? value : null;
   }
 
   /**
-   * Non-empty when {@link #context} is null: A list of additional state extracted from the
-   * request.
+   * Non-empty when {@link #context()} is {@code null}: A list of additional state extracted from
+   * the request.
    *
    * @see TraceContext#extra()
+   * @since 4.9
    */
   public final List<Object> extra() {
-    return extra;
+    return extraList;
   }
 
-  public final Builder toBuilder() {
-    Builder result = new Builder();
-    result.type = type;
-    result.value = value;
-    result.extra = extra;
-    return result;
-  }
-
-  @Override
-  public String toString() {
-    return "{value=" + value + ", extra=" + extra + "}";
-  }
-
-  public static TraceContextOrSamplingFlags create(TraceContext context) {
-    return new TraceContextOrSamplingFlags(1, context, emptyList());
-  }
-
-  public static TraceContextOrSamplingFlags create(TraceIdContext traceIdContext) {
-    return new TraceContextOrSamplingFlags(2, traceIdContext, emptyList());
+  /**
+   * Use to decorate an {@link Extractor#extract extraction result} with {@link #sampledLocal()} or
+   * additional {@link #extra()}.
+   *
+   * @since 4.9
+   */
+  public Builder toBuilder() {
+    return new Builder(type, value, effectiveExtra());
   }
 
-  public static TraceContextOrSamplingFlags create(SamplingFlags flags) {
-    // reuses constants to avoid excess allocation
-    if (flags == SamplingFlags.SAMPLED) return SAMPLED;
-    if (flags == SamplingFlags.EMPTY) return EMPTY;
-    if (flags == SamplingFlags.NOT_SAMPLED) return NOT_SAMPLED;
-    if (flags == SamplingFlags.DEBUG) return DEBUG;
-    return new TraceContextOrSamplingFlags(3, flags, emptyList());
+  @Override public String toString() {
+    List<Object> extra = effectiveExtra();
+    StringBuilder result = new StringBuilder();
+    result.append(value.getClass().getSimpleName()).append('=').append(value);
+    result.setCharAt(0, Character.toLowerCase(result.charAt(0))); // lowercase the first char
```

Comment:
Maybe an extra line or two but seems quite a bit simpler to lowercase before append without the `0`'s
```
result.append("Extracted{")
String valueClass = value.getClass().getSimpleName();
// Lowercase first char of class name
result.append(Character.toLowerCase(valueClass.charAt(0));
result.append(valueClass, 1, valueClass.length());
```

Reply:
thanks good idea